core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java

```diff
@@ -583,8 +588,29 @@ public UpdateNode plan(Update node)
         PlanAndMappings planAndMappings = coerce(builder, orderedColumnValues, analysis, idAllocator, symbolAllocator, typeCoercion);
         builder = planAndMappings.getSubPlan();
 
+        TableMetadata tableMetadata = metadata.getTableMetadata(session, handle);
         ImmutableList.Builder<Symbol> updatedColumnValuesBuilder = ImmutableList.builder();
-        orderedColumnValues.forEach(columnValue -> updatedColumnValuesBuilder.add(planAndMappings.get(columnValue)));
+        Assignments.Builder assignmentsBuilder = Assignments.builder();
+        assignmentsBuilder.putIdentities(builder.getRoot().getOutputSymbols());
+        Map<String, ColumnMetadata> columnNameToMetadata = tableMetadata.getColumns().stream()
+                .collect(toImmutableMap(ColumnMetadata::getName, identity()));
+
+        for (int i = 0; i < targetColumnNames.size(); i++) {
+            String columName = targetColumnNames.get(i);
+            ColumnMetadata columnMetadata = columnNameToMetadata.get(columName);
+            Symbol output = symbolAllocator.newSymbol(columName, columnMetadata.getType());
+            updatedColumnValuesBuilder.add(output);
+            Expression expression = orderedColumnValues.get(i);
+            if (!columnMetadata.isNullable()) {
+                expression = new CoalesceExpression(expression, new Cast(failFunction(metadata, session, INVALID_ARGUMENTS, format(
+                        "NULL value not allowed for NOT NULL column: %s",
+                        columName)), toSqlType(columnMetadata.getType())));
```

Comment:
Format as follows for readability:

```suggestion
                expression = new CoalesceExpression(expression, new Cast(
                        failFunction(
                                metadata,
                                session,
                                INVALID_ARGUMENTS,
                                format("NULL value not allowed for NOT NULL column: %s", columName)),
                        toSqlType(columnMetadata.getType())));
```

Also, you can use string concatenation instead of `format()` for such a simple case:

```
"NULL value not allowed for NOT NULL column: " + columName
```